flink-connector-bigquery/src/test/java/com/google/cloud/flink/bigquery/sink/serializer/AvroToProtoSerializerTest.java

```diff
@@ -16,20 +16,731 @@
+import static com.google.cloud.flink.bigquery.sink.serializer.TestBigQuerySchemas.getAvroSchemaFromFieldString;
+import static com.google.cloud.flink.bigquery.sink.serializer.TestBigQuerySchemas.getRecordSchema;
 import static com.google.common.truth.Truth.assertThat;
+import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.assertThrows;
 
 /** Tests for {@link AvroToProtoSerializer}. */
 public class AvroToProtoSerializerTest {
 
     @Test
-    public void testSerializeIsUnsupported() {
-        UnsupportedOperationException exception =
+    public void testPrimitiveTypesConversion() {
+        BigQuerySchemaProvider bigQuerySchemaProvider =
+                TestBigQuerySchemas.getSchemaWithRequiredPrimitiveTypes();
+        BigQueryProtoSerializer<GenericRecord> serializer = new AvroToProtoSerializer();
+        serializer.init(bigQuerySchemaProvider);
+        Descriptor descriptor = bigQuerySchemaProvider.getDescriptor();
+        Schema avroSchema = bigQuerySchemaProvider.getAvroSchema();
+
+        byte[] byteArray = "Any String you want".getBytes();
+        Schema recordSchema = avroSchema.getField("required_record_field").schema();
+        GenericRecord record =
+                new GenericRecordBuilder(avroSchema)
+                        .set("number", -7099548873856657385L)
+                        .set("price", 0.5616495161359795)
+                        .set("species", "String")
+                        .set("flighted", true)
+                        .set("sound", ByteBuffer.wrap(byteArray))
+                        .set(
+                                "required_record_field",
+                                new GenericRecordBuilder(recordSchema)
+                                        .set("species", "hello")
+                                        .build())
+                        .build();
+
+        DynamicMessage message = getDynamicMessageFromGenericRecord(record, descriptor);
+        assertEquals(-7099548873856657385L, message.getField(descriptor.findFieldByNumber(1)));
+        assertEquals(0.5616495161359795, message.getField(descriptor.findFieldByNumber(2)));
+        assertEquals("String", message.getField(descriptor.findFieldByNumber(3)));
+        assertEquals(true, message.getField(descriptor.findFieldByNumber(4)));
+        assertEquals(
+                ByteString.copyFrom(byteArray), message.getField(descriptor.findFieldByNumber(5)));
+
+        FieldDescriptor fieldDescriptor = descriptor.findFieldByNumber(6);
+        message = (DynamicMessage) message.getField(fieldDescriptor);
+        assertEquals(
+                "hello",
+                message.getField(
+                        descriptor
+                                .findNestedTypeByName(fieldDescriptor.toProto().getTypeName())
+                                .findFieldByNumber(1)));
+    }
+
+    @Test
+    public void testRemainingPrimitiveTypesConversion() {
```

Comment:
let's also add a test for unsupported primitive types (ensuring it throws the right error)

Reply:
Added!